### ISSUE TITLE
invalid nodeSelector "kubernetes.io/hostname", [] not in []

### ISSUE THREAD

**Issue by DiMarcello09** [aws/karpenter#1200]
## Version

**Karpenter**: v0.5.4

**Kubernetes**: v1.21.2-eks-0389ca3

## Expected Behavior
Karpenter should scale up to make PODs schedulable.

## Actual Behavior
PODs are staying in pending state, Karpenter is logging the following:

`2022-01-21T09:01:44.862Z    INFO    controller.provisioning    Unable to schedule <podname>/microservices, invalid nodeSelector "kubernetes.io/hostname", [8eadezcx] not in []    {"commit": "7e79a67", "provisioner":  ││ 2022-01-21T09:01:44.862Z    INFO    controller.provisioning    Unable to schedule pod <podname>/microservices, invalid nodeSelector "kubernetes.io/hostname", [] not in []    {"commit": "7e79a67" ││ 2022-01-21T09:01:44.862Z    INFO    controller.provisioning    Waiting for unschedulable pods    {"commit": "7e79a67", "provisioner": "default"}`

Deployment Spec:

```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "13"
    meta.helm.sh/release-name: app1
    meta.helm.sh/release-namespace: microservices
  creationTimestamp: "2020-11-03T16:25:49Z"
  generation: 16
  labels:
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: app1
  name: app1
  namespace: microservices
  resourceVersion: "191413244"
  uid: 5c056496-1f9a-438e-876f-2490396fbbf1
spec:
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/name: app1
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/config: e962dd0855adc503b1c86a78072956c65edaf2dea9b4626e70c0c30d73f9512e
        kubectl.kubernetes.io/restartedAt: "2021-09-08T15:41:16+02:00"
      creationTimestamp: null
      labels:
        app.kubernetes.io/name: app1
    spec:
      containers:
      - envFrom:
        - configMapRef:
            name: microservices-core
        - configMapRef:
            name: app1
        image: <redacted>
        imagePullPolicy: IfNotPresent
        lifecycle:
          preStop:
            exec:
              command:
              - sleep
              - "40"
        livenessProbe:
          failureThreshold: 3
          httpGet:
            path: /actuator/health/liveness
            port: http
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 15
          successThreshold: 1
          timeoutSeconds: 1
        name: app1
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
        readinessProbe:
          failureThreshold: 2
          httpGet:
            path: /actuator/health/readiness
            port: http
            scheme: HTTP
          initialDelaySeconds: 45
          periodSeconds: 15
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          requests:
            cpu: 150m
            memory: 512Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      nodeSelector:
        karpenter.sh/provisioner-name: default
      readinessGates:
      - conditionType: target-health.alb.ingress.k8s.aws/main_app1_80
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app.kubernetes.io/name: app1
        maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: ScheduleAnyway
      - labelSelector:
          matchLabels:
            app.kubernetes.io/name: app1
        maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: ScheduleAnyway
```

**Comment by ellistarn** [comment]
Hey @DiMarcello09, was this ever working for you? You mentioned in slack that the problem is "getting worse". In the short term, remove the hostname topology rule (if you can), which I think is causing this.

**Comment by ellistarn** [comment]
Hmm -- I'm not able to reproduce this.

```
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.608Z	INFO	controller.provisioning	Batched 2 pods in 1.008748859s	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.616Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [m5.large]	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.879Z	DEBUG	controller.provisioning	Discovered security groups: [sg-000d68e67c3fe53d8]	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.880Z	DEBUG	controller.provisioning	Discovered kubernetes version 1.20	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.941Z	DEBUG	controller.provisioning	Discovered ami ami-07854b27891cd310f for query /aws/service/eks/optimized-ami/1.20/amazon-linux-2/recommended/image_id	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.941Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:33.977Z	DEBUG	controller.provisioning	Discovered launch template Karpenter-etarn-7128528491341215688	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:36.206Z	INFO	controller.provisioning	Launched instance: i-0994c1cac5261ecd2, hostname: ip-192-168-50-71.us-west-2.compute.internal, type: m5.large, zone: us-west-2d, capacityType: spot	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:36.220Z	INFO	controller.provisioning	Bound 1 pod(s) to node ip-192-168-50-71.us-west-2.compute.internal	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:36.228Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 1 pod(s) with instance type option(s) [m5.large]	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:36.228Z	DEBUG	controller.provisioning	Discovered caBundle, length 1025	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:38.355Z	INFO	controller.provisioning	Launched instance: i-0c62649d9dd7f94a2, hostname: ip-192-168-81-192.us-west-2.compute.internal, type: m5.large, zone: us-west-2a, capacityType: spot	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:38.370Z	INFO	controller.provisioning	Bound 1 pod(s) to node ip-192-168-81-192.us-west-2.compute.internal	{"commit": "4076128", "provisioner": "default"}
karpenter-controller-6dbb4b47cf-gsrk8 manager 2022-01-21T17:37:38.370Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "4076128", "provisioner": "default"}
```
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: app1
  name: app1
spec:
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/name: app1
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        checksum/config: e962dd0855adc503b1c86a78072956c65edaf2dea9b4626e70c0c30d73f9512e
        kubectl.kubernetes.io/restartedAt: "2021-09-08T15:41:16+02:00"
      labels:
        app.kubernetes.io/name: app1
    spec:
      containers:
      - image: public.ecr.aws/eks-distro/kubernetes/pause:3.2
        imagePullPolicy: IfNotPresent
        lifecycle:
          preStop:
            exec:
              command:
              - sleep
              - "40"
        name: app1
        ports:
        - containerPort: 8080
          name: http
          protocol: TCP
        resources:
          requests:
            cpu: 150m
            memory: 512Mi
      dnsPolicy: ClusterFirst
      nodeSelector:
        karpenter.sh/provisioner-name: default
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app.kubernetes.io/name: app1
        maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: ScheduleAnyway
      - labelSelector:
          matchLabels:
            app.kubernetes.io/name: app1
        maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: ScheduleAnyway
```

**Comment by ellistarn** [comment]
Can you verify your karpenter version for me?

**Comment by DiMarcello09** [comment]
So on the initial rollout this worked absolutely fine and right now its not anymore.

Verifying karpenter version:

`kubectl get pods -n karpenter -l karpenter=controller -o jsonpath="{.items[*].spec.containers[*].image}"
public.ecr.aws/karpenter/controller:v0.5.4@sha256:19ebf83d64fa41d75fb19ae5047f54b1c66423b4ab5ceef36ae99a0daaad1895`

**Comment by DiMarcello09** [comment]
I removed the hostname topology on one deployment spec, Karpenter scaled up again and the other PODs were also able getting scheduled now (where I haven't removed the hostname topology yet).

**Comment by ellistarn** [comment]
This is very surprising. Can you post as many logs as you have from the karpenter controller?

**Comment by prateekgogia** [comment]
For reference I am also on the latest (v0.5.5) Karpenter version and the following specs work for me-
Provisioner spec-
```
spec:
  kubeletConfiguration: {}
  labels:
    kit.k8s.sh/app: example-apiserver
    kit.k8s.sh/control-plane-name: example
  limits: {}
  provider:
    apiVersion: extensions.karpenter.sh/v1alpha1
    instanceProfile: KarpenterNodeInstanceProfile-new-cluster
    kind: AWS
    securityGroupSelector:
      kubernetes.io/cluster/new-cluster: '*'
    subnetSelector:
      kubernetes.io/cluster/new-cluster: '*'
  requirements:
  - key: kubernetes.io/arch
    operator: In
    values:
    - amd64
  - key: karpenter.sh/capacity-type
    operator: In
    values:
    - on-demand
  ttlSecondsAfterEmpty: 30
```

TopologySpread on Pod -
```
 topologySpreadConstraints:
  - labelSelector:
      matchLabels:
        kit.k8s.sh/app: example-apiserver
    maxSkew: 1
    topologyKey: topology.kubernetes.io/zone
    whenUnsatisfiable: DoNotSchedule
  - labelSelector:
      matchLabels:
        kit.k8s.sh/app: example-apiserver
    maxSkew: 1
    topologyKey: kubernetes.io/hostname
    whenUnsatisfiable: DoNotSchedule
```

**Comment by ellistarn** [comment]
Any ideas for differences I may not be accounting for in my reproduction? Perhaps a provisioner spec?

**Comment by DiMarcello09** [comment]
Here is the log of the last like 50 minutes. I had debugging mode enabled, if I did it right.

[karpenter.log](https://github.com/aws/karpenter/files/7915735/karpenter.log)

**Comment by ellistarn** [comment]
This is very strange to me. I'm seeing hostname values that match the hashing strategy we use for hostname topology, but it's not clear to me why this is happening. Let me deep dive to code.

**Comment by DiMarcello09** [comment]
Any news on this? It happened again today. I haven't removed the hostname topology yet.

**Comment by ellistarn** [comment]
Hey @DiMarcello09. I've marked this as a burning bug. We will get to the bottom of this as soon as we can. Can you provide a provisioner spec?

**Comment by DiMarcello09** [comment]
Provisioner spec we're using is the following:

```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  requirements:
    - key: "karpenter.sh/capacity-type"
      operator: In
      values: ["spot"]
    - key: "node.kubernetes.io/instance-type"
      operator: In
      values: ["a1.xlarge", "c5.xlarge", "c5a.xlarge", "c5ad.xlarge", "c5d.xlarge", "c6g.xlarge", "c6gd.xlarge", "inf1.xlarge"]
  limits:
    resources:
      cpu: 1000
  provider:
    instanceProfile: ${instance_profile}
    subnetSelector:
      kubernetes.io/role/internal-elb: "1"
  ttlSecondsAfterEmpty: 300
```

**Comment by kgyovai** [comment]
__Kubernetes: v1.21.2-eks-06eac09__

I'm encountering what seems to be an issue with `topologyKey: kubernetes.io/hostname` that is affecting scheduling but my particular issue is occurring when I'm attempting to use `podAntiAffinity` rather than `topologySpreadConstraints`.

```yaml
    spec:
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            - labelSelector:
                matchLabels:
                  app.kubernetes.io/name: argocd-redis-ha
              topologyKey: kubernetes.io/hostname
```

Expected: 3 replicas within stateful set are scheduled 1 per node (3 node EKS cluster)

Actual: Pod is pending
NAME                       READY   STATUS    RESTARTS   AGE
argocd-redis-ha-server-0   2/2     Running   0          13h
argocd-redis-ha-server-1   0/2     Pending   0          13h

This is happening for any `podAntiAffinity` that attempts to use `topologyKey: kubernetes.io/hostname` even though each node is correctly labeled.

NAME                                        STATUS   ROLES    AGE   VERSION
ip-10-20-1-68.us-east-2.compute.internal    Ready    <none>   21d   v1.21.5-eks-bc4871b
ip-10-20-2-207.us-east-2.compute.internal   Ready    <none>   37h   v1.21.5-eks-bc4871b
ip-10-20-3-211.us-east-2.compute.internal   Ready    <none>   21d   v1.21.5-eks-bc4871b

![image](https://user-images.githubusercontent.com/10728023/151161394-3f208410-ffe5-4065-a7c5-5035a98c0ce9.png)

**Comment by ellistarn** [comment]
Ok I have an idea -- it's really nuanced and tricky.

We currently cache pod.spec.affinity in order to implement preference relaxation. If we see a pod a second time, we attempt to remove preference based requirements in the hope that it will be able to schedule (e.g. spot prioritization). 

Elsewhere in the code, the hostname topology implementation injects a fake hostname into a pod in order for the scheduling logic to separate it from other pods. However, I neglected to realize that since this is using a pointer, it will actually update the cached affinity. If this pod then comes back around, it'll get discovered in the affinity cache, and fail to schedule due to the bad hostname label. 

This is one of the rare places in the codebase that violates immutability principles, and has bitten us before :(

This may not be your specific issue, but it definitely is *an* issue. Let me get a fix out for this.

**Comment by ellistarn** [comment]
This hypothesis was wrong. The injection logic only uses NodeSelectors, which aren't included in affinity.

**Comment by ellistarn** [comment]
@kgyovai, your issue is separate. PodAntiaffinity is not supported (yet). Feel free to +1 https://github.com/aws/karpenter/issues/942

**Comment by kgyovai** [comment]
@ellistarn - You're right. This wasn't karpenter related. My mistake.

This ended up being 100% self inflicted. The AMI that I was using is created from a base AMI. In the base AMI, I had configured preserve hostname. So when the downstream instances were launched using that AMI, the hostname of the instance used to create the AMI was being recorded as the `kubernetes.io/hostname`. Once I dropped hostname preservation the affinity rules worked as expected.
